misc/python/materialize/cloudtest/util/authentication.py

```diff
@@ -20,22 +21,29 @@
 
 @dataclass
 class AuthConfig:
-    organization_id: str
-    token: str
-    app_password: Optional[str]
+    organization_id: str = field(init=False)
+    token: str = field(init=False)
+    app_user: Optional[str] = field(init=False)
+    app_password: Optional[str] = field(init=False)
+
+    refresh_fn: Callable[[AuthConfig], None]
+
+    client_cert: Optional[tuple[str, str]] = None
+
+    def refresh(self) -> None:
+        self.refresh_fn(self)
 
 
 DEFAULT_ORG_ID = "80b1a04a-2277-11ed-a1ce-5405dbb9e0f7"
 
 # TODO: this retry loop should not be necessary, but we are seeing
 # connections getting frequently (but sporadically) interrupted here - we
 # should track this down and remove these retries
-def get_auth(args: argparse.Namespace) -> AuthConfig:
-    config: AuthConfig = retry(lambda: _get_auth(args), 5, [ConnectionError])
-    return config
+def update_auth(args: argparse.Namespace, config: AuthConfig) -> None:
```

Comment:
Is there a reason you switched this to require an `AuthConfig` that will be updated/modified rather than returning one?

Generally I don't love dataclasses with required but non-initialized fields. Since they throw exceptions if the dataclass instance is used anywhere without all the required fields populated.

Reply:
I am also not to happy that I had to do this change.

The reason is the `web_requests.py` file. It takes `AuthConfig` as input param, uses it, but if it fails with 401 it will refresh the authentication and retry. So, I can either update `AuthConfig` to account for that (so that future calls won't run into the same and also require a re-auth) or return instead of the `Response` a tuple of `Response` and original or new authentication.

Reply:
What I could do is to encapsulate the request functions as well in a class and store the recent `AuthConfig` there... I think this would allow to clean this up.

Reply:
>The reason is the web_requests.py file. It takes AuthConfig as input param, uses it, but if it fails with 401 it will refresh the authentication and retry. So, I can either update AuthConfig to account for that (so that future calls won't run into the same and also require a re-auth) or return instead of the Response a tuple of Response and original or new authentication.

I'm not sure that I grok this. When the authentication is refreshed on a 401 error, can't the `refresh_fn` edit the fields on the `AuthConfig` object directly since you are providing `self` as an argument to that function? We aren't calling this `update_auth` function from anything in `web_requests.py` so I'm a little confused on how it's related

Reply:
Oh, unless you're expecting that most of the tests will use `update_auth` as their `refresh_fn`. Maybe we could make this optionally take an `AuthConfig`, but if it's not provided then it can create a fresh one. And if it is provided it will update it. But we can remove the `init=False` from the required fields